toolchain/check/function.cpp

```diff
@@ -191,60 +191,91 @@ auto CheckFunctionTypeMatches(Context& context,
+                         SemIR::FunctionId prev_function_id,
+                         bool prev_is_import) -> bool {
   auto& prev_function = context.functions().Get(prev_function_id);
 
   if (!CheckRedecl(context, new_function, prev_function, {})) {
     return false;
   }
 
-  if (!is_definition) {
-    CARBON_DIAGNOSTIC(FunctionRedecl, Error,
-                      "Redundant redeclaration of function {0}.",
-                      SemIR::NameId);
-    CARBON_DIAGNOSTIC(FunctionPreviousDecl, Note, "Previously declared here.");
-    context.emitter()
-        .Build(node_id, FunctionRedecl, prev_function.name_id)
-        .Note(prev_function.decl_id, FunctionPreviousDecl)
-        .Emit();
-    // The diagnostic doesn't prevent a merge.
-    return true;
-  } else if (prev_function.definition_id.is_valid()) {
-    CARBON_DIAGNOSTIC(FunctionRedefinition, Error,
-                      "Redefinition of function {0}.", SemIR::NameId);
-    CARBON_DIAGNOSTIC(FunctionPreviousDefinition, Note,
-                      "Previously defined here.");
-    context.emitter()
-        .Build(node_id, FunctionRedefinition, prev_function.name_id)
-        .Note(prev_function.definition_id, FunctionPreviousDefinition)
-        .Emit();
-    // The second definition will be unused as a consequence of the error.
-    return true;
-  } else if (prev_function.is_extern) {
-    CARBON_DIAGNOSTIC(FunctionDefiningExtern, Error,
-                      "Cannot define `extern` function `{0}`.", SemIR::NameId);
-    CARBON_DIAGNOSTIC(FunctionPreviousExternDecl, Note,
-                      "Previously declared `extern` here.");
-    context.emitter()
-        .Build(node_id, FunctionDefiningExtern, prev_function.name_id)
-        .Note(prev_function.decl_id, FunctionPreviousExternDecl)
-        .Emit();
-    // The diagnostic doesn't prevent a merge.
-    return true;
-  }
+  CheckIsAllowedRedecl(context, node_id, new_function, new_is_definition,
+                       prev_function, prev_is_import);
 
-  // Track the signature from the definition, so that IDs in the body
-  // match IDs in the signature.
-  prev_function.definition_id = new_function.definition_id;
-  prev_function.implicit_param_refs_id = new_function.implicit_param_refs_id;
-  prev_function.param_refs_id = new_function.param_refs_id;
-  prev_function.return_type_id = new_function.return_type_id;
-  prev_function.return_slot_id = new_function.return_slot_id;
+  if (new_is_definition) {
+    // Track the signature from the definition, so that IDs in the body
+    // match IDs in the signature.
+    prev_function.definition_id = new_function.definition_id;
+    prev_function.implicit_param_refs_id = new_function.implicit_param_refs_id;
+    prev_function.param_refs_id = new_function.param_refs_id;
+    prev_function.return_type_id = new_function.return_type_id;
+    prev_function.return_slot_id = new_function.return_slot_id;
+  }
+  if (!new_function.is_extern) {
+    prev_function.is_extern = false;
+  }
```

Comment:
What happens for:
```carbon
package A library "foo" api;
fn F();
```
```carbon
package A library "bar" api;
// 1
import library "foo";
// 2
extern fn F();
// 3
fn F() {}
```
To my reading of the code:
At (1), we import a non-`extern` declaration of `F`. `is_extern` is `false`.
At (2), we process a valid redeclaration, and don't change `is_extern` here.
At (3), we process a non-`extern` definition, the previous declaration is not imported, and the function is not extern, so we allow the definition.

Perhaps:
```suggestion
  // The first non-imported declaration determines whether the function is extern in this file. For error recovery, if we see both an extern declaration and a non-extern declaration in the same file, consider the function to be non-extern.
  if (prev_is_import || !new_function.is_extern) {
    prev_function.is_extern = new_function.is_extern;
  }
```

Reply:
The suggested prev_is_import won't be the logic that's desired long-term, because you can also get cases where the previous is an import and the current case is *also* an import. You would be fixing one case, but I believe it introduces other bugs.

To be sure, there is a TODO at head for this, and this change leaves that TODO in place.

```
// TODO: Detect conflicting cross-file declarations, as well as uses of imported
// declarations followed by a redeclaration.
```

https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/check/function.cpp#L194-L195

Reply:
Also, adding fail_mix_extern_decl.carbon to demonstrate behavior.

Reply:
Has that been pushed yet? I can't find the test file in this PR.

Reply:
It's a split-file in the final commit:

https://github.com/carbon-language/carbon-lang/pull/3809/commits/3b047acfd09a2c7f684849d87ff2928ef8d28067#diff-5637904bada28e29921d8cf1c93f584d73ef1fe9e5c087cff42a5852d68ecc02